rule correct_error_paired_reads_with_spades:
    threads:
        16
    conda:
        pipeline_path + "envs/spades.yaml"
    input:
        R1_trimmed = "samples/{sample}/reads/trimmed/R1_paired.fastq",
        R2_trimmed = "samples/{sample}/reads/trimmed/R2_paired.fastq",
    output:
        R1_corrected = temp("samples/{sample}/reads/corrected/R1_paired.00.0_0.cor.fastq.gz"),
        R2_corrected = temp("samples/{sample}/reads/corrected/R2_paired.00.0_0.cor.fastq.gz"),
        single_corrected = temp("samples/{sample}/reads/corrected/R_unpaired.00.0_0.cor.fastq.gz"),
    log:
        logging_folder + "samples/{sample}/logs/spades_read_correction.txt"
    shell:
         """
         spades.py --only-error-correction -1 {input[R1_trimmed]} -2 {input[R2_trimmed]} -o $(dirname $(dirname {output[0]})) --threads {threads} > {log} 
         """

rule correct_error_single_reads_with_spades:
    threads:
        16
    conda:
        pipeline_path + "envs/spades.yaml"
    input:
        single_trimmed = "samples/{sample}/reads/trimmed/single.fastq",
    output:
        single_corrected = temp("samples/{sample}/reads/corrected/single.00.0_0.cor.fastq.gz")
    log:
        logging_folder + "samples/{sample}/logs/spades_read_correction.txt"
    shell:
         """
         spades.py --only-error-correction  -s {input[single_trimmed]} -o $( dirname $(dirname {output[single_corrected]})) --threads {threads} > {log} 
         """
         
if config['meta'] != True:
         
    rule assemble_genome_corrected_paired_reads_with_spades:
        threads:
            16
        conda:
            pipeline_path + "envs/spades.yaml"
        input:
            R1_corrected = "samples/{sample}/reads/corrected/R1_paired.00.0_0.cor.fastq.gz",
            R2_corrected = "samples/{sample}/reads/corrected/R2_paired.00.0_0.cor.fastq.gz",
            single_corrected = "samples/{sample}/reads/corrected/R_unpaired.00.0_0.cor.fastq.gz",
        output:
            contigs = "samples/{sample}/assembly/spades/contigs.fasta",
            graph = "samples/{sample}/assembly/spades/assembly_graph.gfa",
        log:
            logging_folder + "samples/{sample}/logs/spades_assembly.txt"
        shell:
            """
        spades.py --only-assembler -1 {input[R1_corrected]} -2 {input[R2_corrected]} -s {input[single_corrected]} -o $( dirname {output[contigs]}) --threads {threads} > {log}
        mv $(dirname {output[contigs]})/contigs.fasta $(dirname $(dirname {output[contigs]}))
        mv $(dirname {output[contigs]})/assembly_graph.gfa $(dirname $(dirname {output[contigs]}))
        rm -rf $( dirname {output[contigs]})/*
        mv $(dirname $(dirname {output[contigs]}))/contigs.fasta $(dirname {output[contigs]})/
        mv $(dirname $(dirname {output[contigs]}))/assembly_graph.gfa $(dirname {output[contigs]})/
            """
else:
        
    rule assemble_metagenome_corrected_paired_reads_with_spades:
        threads:
            16
        conda:
            pipeline_path + "envs/spades.yaml"
        input:
            R1_corrected = "samples/{sample}/reads/corrected/R1_paired.00.0_0.cor.fastq.gz",
            R2_corrected = "samples/{sample}/reads/corrected/R2_paired.00.0_0.cor.fastq.gz",
            single_corrected = "samples/{sample}/reads/corrected/R_unpaired.00.0_0.cor.fastq.gz",
        output:
            contigs = "samples/{sample}/assembly/spades/contigs.fasta",
            graph = "samples/{sample}/assembly/spades/assembly_graph.fastg",
            scaffolds = "samples/{sample}/assembly/spades/scaffolds.fasta",
            contigs_paths = "samples/{sample}/assembly/spades/contigs.paths",
            scaffolds_paths = "samples/{sample}/assembly/spades/scaffolds.paths",
        log:
            logging_folder + "samples/{sample}/logs/spades_assembly.txt"
        shell:
            """
            #spades.py --only-assembler -k 127 --meta -1 {input[R1_corrected]} -2 {input[R2_corrected]} -s {input[single_corrected]} -o $( dirname {output[contigs]}) --threads {threads} > {log}
            # move to previous dir    
            mv $(dirname {output[contigs]})/contigs.fasta $(dirname $(dirname {output[contigs]}))
            mv $(dirname {output[contigs]})/assembly_graph.fastg $(dirname $(dirname {output[contigs]}))
            mv $(dirname {output[contigs]})/scaffolds.fasta $(dirname $(dirname {output[contigs]}))    
            mv $(dirname {output[contigs]})/contigs.paths $(dirname $(dirname {output[contigs]}))
            mv $(dirname {output[contigs]})/scaffolds.paths $(dirname $(dirname {output[contigs]}))
            # remove unnessesary files
            echo "rm -rf $( dirname {output[contigs]})/*"
            #rm -rf $( dirname {output[contigs]})/*
            # move back to spades directory    
            mv $(dirname $(dirname {output[contigs]}))/contigs.fasta $(dirname {output[contigs]})/
            mv $(dirname $(dirname {output[contigs]}))/assembly_graph.fastg $(dirname {output[contigs]})/
            mv $(dirname $(dirname {output[contigs]}))/scaffolds.fasta $(dirname {output[contigs]})/
            mv $(dirname $(dirname {output[contigs]}))/assembly_graph.fastg $(dirname {output[contigs]})/    
            """

rule assemble_genome_corrected_single_reads_with_spades:
    threads:
        16
    conda:
        pipeline_path + "envs/spades.yaml"
    input:
        single_corrected = "samples/{sample}/reads/corrected/single.00.0_0.cor.fastq.gz",
    output:
        contigs = "samples/{sample}/assembly/spades/contigs.fasta"
    log:
        logging_folder + "samples/{sample}/logs/spades_assembly.txt",
    shell:
        """
        spades.py --only-assembler -s {input[single_corrected]} -o $( dirname {output[contigs]}) --threads {threads} > {log}
        mv $(dirname {output[contigs]})/contigs.fasta $(dirname $(dirname {output[contigs]}))
        rm -rf $( dirname {output[contigs]})/*
        mv $(dirname $(dirname {output[contigs]}))/contigs.fasta $(dirname {output[contigs]})/
        """
        
rule assemble_genome_paired_reads_with_spades:
    threads:
        16
    conda:
        pipeline_path + "envs/spades.yaml"
    input:
        R1_trimmed = "samples/{sample}/reads/trimmed/R1_paired.fastq",
        R2_trimmed = "samples/{sample}/reads/trimmed/R2_paired.fastq",
    output:
        "samples/{sample}/assembly/spades_no_correction/contigs.fasta"
    log:
        logging_folder + "samples/{sample}/logs/spades_no_correction_assembly.txt"
    shell:
        """
        spades.py --only-assembler -1 {input[R1_trimmed]} -2 {input[R2_trimmed]} -o $( dirname {output[0]}) > {log}
        mv $(dirname {output[0]})/contigs.fasta $(dirname $(dirname {output[0]}))
        rm -rf $( dirname {output[0]})/*
        mv $(dirname $(dirname {output[0]}))/contigs.fasta $(dirname {output[0]})/
        """

rule assemble_genome_single_reads_with_spades:
    threads:
        16
    conda:
        pipeline_path + "envs/spades.yaml"
    input:
        single_trimmed = "samples/{sample}/reads/trimmed/single.fastq",
    output:
        "samples/{sample}/assembly/spades_no_correction/contigs.fasta"
    log:
        logging_folder + "samples/{sample}/logs/spades_no_correction_assembly.txt",
    shell:
        """
        spades.py --only-assembler -s {input[single_trimmed]} -o $( dirname {output[0]}) > {log}
        mv $(dirname {output[0]})/contigs.fasta $(dirname $(dirname {output[0]}))
        rm -rf $( dirname {output[0]})/*
        mv $(dirname $(dirname {output[0]}))/contigs.fasta $(dirname {output[0]})/
        """
        
