<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Logging functions &#8212; diag_pipelines  documentation</title>
    
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Workflows" href="workflows.html" />
    <link rel="prev" title="Documentation for the genomics and metagenomics workflows" href="index.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="logging-functions">
<h1>Logging functions<a class="headerlink" href="#logging-functions" title="Permalink to this headline">¶</a></h1>
<p>Archiving processes are defined in the file <code class="file docutils literal"><span class="pre">../workflows/logging.rules</span></code>. The variable <code class="docutils literal"><span class="pre">logging_folder</span></code> must be defined in the <code class="docutils literal"><span class="pre">config.yaml</span></code> or passed to snakemake with <code class="docutils literal"><span class="pre">--config</span></code>. Each time an effective snakemake run is started, a folder named with the current UTC datetime. A different number of files will be copied there, so that replication of the run is possible:</p>
<ul class="simple">
<li>The snakefile passed to snakemake</li>
<li>The config file</li>
<li>The full command used, copied into the file <code class="docutils literal"><span class="pre">cmd.txt</span></code></li>
<li>The parameter files defined the SRA and the local samples, if they exist</li>
</ul>
</div>
<div class="section" id="determining-sample-names">
<h1>Determining sample names<a class="headerlink" href="#determining-sample-names" title="Permalink to this headline">¶</a></h1>
<p>Samples for the run will be determined in the file <code class="file docutils literal"><span class="pre">../workflows/making_sample_dataset.rules</span></code>.</p>
<div class="section" id="local-samples">
<h2>Local samples<a class="headerlink" href="#local-samples" title="Permalink to this headline">¶</a></h2>
<p>Local samples will be determined based on the tabulated file whose full path must be passed to the variable <code class="docutils literal"><span class="pre">local_samples</span></code> in the <code class="docutils literal"><span class="pre">config.yaml</span></code> or through <code class="docutils literal"><span class="pre">--config</span></code> on the snakemake command. It must contain at least two columns: <cite>SampleName</cite> and <cite>ScientificName</cite>.</p>
<table border="1" class="docutils" id="id1">
<caption><span class="caption-text">Local data example</span><a class="headerlink" href="#id1" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="50%" />
<col width="50%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">SampleName</th>
<th class="head">ScientificName</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>S10</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-odd"><td>S1</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-even"><td>S2</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-odd"><td>S3</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-even"><td>S4</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-odd"><td>S5</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-even"><td>S6</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-odd"><td>S7</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-even"><td>S8</td>
<td>Staphylococcus aureus</td>
</tr>
<tr class="row-odd"><td>S9</td>
<td>Staphylococcus aureus</td>
</tr>
</tbody>
</table>
<p>For each entry, there must be in the folder defined in variable <code class="docutils literal"><span class="pre">link_directory</span></code>, two files (for paired reads) or a single one (for single reads) whose filename starts with one and only one entry in the <cite>SampleName</cite> columns. For instance, the files <code class="docutils literal"><span class="pre">S10_001_R1_L001.fastq.gz</span></code> and <code class="docutils literal"><span class="pre">S10_001_R2_L001.fastq.gz</span></code> in the folder defined by <code class="docutils literal"><span class="pre">link_directory</span></code> folder will be matched to the sample name <code class="docutils literal"><span class="pre">S10</span></code>.</p>
<p>If needed, an <cite>OldSampleName</cite> column can be added to the file, when the read file names and the desired new sample names can not be match simply by testing the identity at the start of both strings.</p>
<table border="1" class="docutils" id="id2">
<caption><span class="caption-text">Local data example with old sample names</span><a class="headerlink" href="#id2" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="33%" />
<col width="33%" />
<col width="33%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">SampleName</th>
<th class="head">ScientificName</th>
<th class="head">OldSampleName</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>S10</td>
<td>Staphylococcus aureus</td>
<td>Staaur-10</td>
</tr>
<tr class="row-odd"><td>S1</td>
<td>Staphylococcus aureus</td>
<td>Staaur-1</td>
</tr>
<tr class="row-even"><td>S2</td>
<td>Staphylococcus aureus</td>
<td>Staaur-2</td>
</tr>
<tr class="row-odd"><td>S3</td>
<td>Staphylococcus aureus</td>
<td>Staaur-3</td>
</tr>
<tr class="row-even"><td>S4</td>
<td>Staphylococcus aureus</td>
<td>Staaur-4</td>
</tr>
<tr class="row-odd"><td>S5</td>
<td>Staphylococcus aureus</td>
<td>Staaur-5</td>
</tr>
<tr class="row-even"><td>S6</td>
<td>Staphylococcus aureus</td>
<td>Staaur-6</td>
</tr>
<tr class="row-odd"><td>S7</td>
<td>Staphylococcus aureus</td>
<td>Staaur-7</td>
</tr>
<tr class="row-even"><td>S8</td>
<td>Staphylococcus aureus</td>
<td>Staaur-8</td>
</tr>
<tr class="row-odd"><td>S9</td>
<td>Staphylococcus aureus</td>
<td>Staaur-9</td>
</tr>
</tbody>
</table>
<p>In this case, the files <code class="docutils literal"><span class="pre">Staaur-10_S10_L001_R1_001.fastq.gz</span></code> and <code class="docutils literal"><span class="pre">Staaur-10_S10_L001_R2_001.fastq.gz</span></code> in the folder defined in <code class="docutils literal"><span class="pre">link_directory</span></code> will be matched to the sample name <code class="docutils literal"><span class="pre">S10</span></code>.</p>
</div>
<div class="section" id="sra-samples">
<h2>SRA samples<a class="headerlink" href="#sra-samples" title="Permalink to this headline">¶</a></h2>
<p>SRA samples will be determined based on the tabulated file whose full path must be passed to the variable <code class="docutils literal"><span class="pre">sra_samples</span></code>. The RunInfo files that can be downloaded through the NCBI can be directly passed without any modification. Otherwise, four columns must be defined.</p>
<table border="1" class="docutils" id="id3">
<caption><span class="caption-text">SRA data example</span><a class="headerlink" href="#id3" title="Permalink to this table">¶</a></caption>
<colgroup>
<col width="25%" />
<col width="25%" />
<col width="25%" />
<col width="25%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Run</th>
<th class="head">SampleName</th>
<th class="head">LibraryLayout</th>
<th class="head">ScientificName</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>ERR1140788</td>
<td>Mycobacterium_tuberculosis_N0145-Lineage_2</td>
<td>paired</td>
<td>Mycobacterium tuberculosis</td>
</tr>
<tr class="row-odd"><td>SRR006916</td>
<td>Mycobacterium_tuberculosis_K21-Lineage_1</td>
<td>single</td>
<td>Mycobacterium tuberculosis</td>
</tr>
<tr class="row-even"><td>SRR022880</td>
<td>Mycobacterium_tuberculosis_africanum_4141_04-Lineage_6</td>
<td>single</td>
<td>Mycobacterium tuberculosis</td>
</tr>
<tr class="row-odd"><td>ERR551336</td>
<td>Mycobacterium_tuberculosis_11821-03-Lineage_5</td>
<td>paired</td>
<td>Mycobacterium tuberculosis</td>
</tr>
<tr class="row-even"><td>ERR756345</td>
<td>Mycobacterium_tuberculosis_aethiop_vetus-Lineage_7</td>
<td>paired</td>
<td>Mycobacterium tuberculosis</td>
</tr>
<tr class="row-odd"><td>ERR238745</td>
<td>Mycobacterium_tuberculosis_SG1-Lineage_3</td>
<td>single</td>
<td>Mycobacterium tuberculosis</td>
</tr>
</tbody>
</table>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Logging functions</a></li>
<li><a class="reference internal" href="#determining-sample-names">Determining sample names</a><ul>
<li><a class="reference internal" href="#local-samples">Local samples</a></li>
<li><a class="reference internal" href="#sra-samples">SRA samples</a></li>
</ul>
</li>
</ul>
<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Documentation for the genomics and metagenomics workflows</a></li>
      <li>Next: <a href="workflows.html" title="next chapter">Workflows</a></li>
  </ul></li>
</ul>
</div>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/general_processes.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Sacha Laurent.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.5.6</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/general_processes.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>